### ISSUE TITLE
Tracks on display slow Osmand down (configure map > tracks is terribly slow too)

### ISSUE THREAD

**Issue by Rowin63** [osmandapp/OsmAnd#20067]
### Description

One thing is, since some time (some weeks) Osmand slows down terribly as soon as there are some tracks on display. Getting worse by every displayed track. 
The other thing is, the tracks section ("configure maps > tracks") takes about 1m20s until the content is displayed (this does not happen via "my places > tracks"!)

### Steps to reproduce

I don´t really know, but it obviously is caused by the number of stored/displayed tracks. 
Without tracks the app "configure map > tracks" runs smooth.
Without tracks on display the app itself (zoom/pan the map) runs smooth.

### Actual result

I did several fresh installs the last days and tried different situations (changing number of maps/favorites/tracks/contour lines/wiki/overlay map/... contained in the apps folders).
A) As long as there are no tracks on display, the app runs smooth. 
As soon as I swith the first few traclks on display, Osmand slows down. 

B) As long as there are no tracks stored in the app, "configure map > tracks" opens quick
As soon as I store tracks in the app, "configure map > tracks" gets slower and slower by the number of tracks

The only difference I see: 
this happens only on my Samsung Tab S6 running Android 12 (since some weeks) but not on my two other devices running Android 13 (Tab Act3, S20+).

I use androids "media"-folder, but this problem also exists in "data" ("extern 1")

But anyway it´s a "new" behaviour and did not happen before. 
Data is the same on all devices and far away from the previous "oom"-numbers.

### Expected result

Don´t slow down like this...

### Your Environment (required)

OsmAnd Version: +4.7.17
Android/iOS version: 12 (Tab S6), 13 (Tab Act, S20+)

**Comment by yuriiurshuliak** [comment]
The bug has been reproduced. Initially, zooming and scrolling the map without displaying the tracks worked smoothly and consistently. However, once I enabled the display of 47 tracks, the map began to slow down and lag when zooming and moving around.

OsmAnd~ 4.8.0#2476m, released: 2024-06-10

<table style="width:100%">
  </tr>
  <tr>
    <th>No tracks on the map</th>
    <th>With tracks on the map</th>
  </tr>
  <tr>
    <td> <video source src="https://github.com/osmandapp/OsmAnd/assets/127092082/1b07fa5d-9198-423f-9bf8-1a22c4fdeff1"  width="300" height="auto"/>   </td>
    <td><video source src="https://github.com/osmandapp/OsmAnd/assets/127092082/8dc1c6b6-57dd-42cc-a2c7-a18683fff46c"  width="300" height="auto"/>   </td>
  </tr>
</table>

**Comment by Rowin63** [comment]
Thanks for investigating and confirming 👍
Please tell me, do you see this related to Android 12?

**Comment by sonora** [comment]
In my experience, it is a weakness of the v2 engine. I am using `Menu > Settings > OsmAnd settings > Map rendering engine > Version 1`, with much better performance (Android 11). May depend on device, though.

**Comment by Rowin63** [comment]
I already switched back to V1, does not make a difference in this issue.
The main problem is this looong time Osmand needs to display the tracks in "configure map > tracks". As mentionend, it takes much more than one minute until the tracks are processed/displayed. 
But now I see, in fact this also happens somehow in "my places > tracks". Here at least the folders are displayed almost immediately, but also here it takes a lot of time until all the tracks are processed and displayed.

Strange thing is, this does not happen at all on my other (android 13) devices. Everything is displayed within a few seconds.
And the Tab S6 should be better hardware than the Tab Act3 is. Though it does not happen on the Act3.
So the only dfference I can see is A12 and A13. Dont know what else to do...

And I did really clean/fresh installs (deleting all remaining folders/data via file manager after de-installing...)

**Comment by sonora** [comment]
Which build number exactly? I had a similar issue and have done a lot of work in the last 3 weeks or so on that, finally bringing things down to 1-3 sec for either screen (for my setup with~4000 tracks on 100 folders). But it depends a bit on the exact build, because we had some interim commits which had brought back the "slowness"... (and I think the public release is still missing this optimization entirely.)

In any case I think optimizing the code for these two screens has now reached its limit. There are 2 other epics on the way optimizing the Track loading and caching in general, they are still work in progress.

**Comment by Rowin63** [comment]
I have to reduce the content of the tracks folder to a REAL minimum (120 tracks, less than 50Mb), then the app runs smooth and the tracks list is displayed within a few seconds. No matter if 5 or 50 maps downloaded, it's just the amount of stored tracks.

**Comment by Rowin63** [comment]
It's the +4.7.17 from playstore.

**Comment by sonora** [comment]
The things I fixed was avoiding repeated file I/O, and removing recursive code. While the first impacts most devices, the second point depends strongly on how much memory a device has and how it allocates/reuses it during excessive recursion. So quite possible that you observe such difference between devices, and also quite possible that if you test a nightly build or wait for the next release, you could see a decisive improvement. :wink: I for my setup saw an overall improvement of a factor of 10-15...

**Comment by Rowin63** [comment]
Just to countercheck this issue.
I organized another Tab S6 (same SM-T865, same Android 12), installed the same nightly and it shows the same issue with readig/processing the tracks.
So it can be caused by the Tab S6 itself (but it's an Octacore, 8Gb/256Gb) , or the Android 12.
I'm still confused, not an attempt of this issue on my Android 13-devices...

**Comment by sonora** [comment]
Does it make a performance difference if you use external or internal storage for the OsmAnd data folder?

**Comment by Rowin63** [comment]
Have to check this.
I'm using the "media"-path (for access/synching reasons), and I've tried the "obb" -path, no difference.

Next I'll give "internal" a try.

**Comment by Rowin63** [comment]
> Does it make a performance difference if you use external or internal storage for the OsmAnd data folder?

Pfffthhh...
Using the "internal" path changes Osmand into a rocket...
Tracks/content ist displayed/processed in the fraction of a second. You can't even notice a delay...
This shows this issues cause, but I can't use the internal memory, it's hidden and I don't have access to the data per file manager...

**Comment by sonora** [comment]
Ok, that confirms my suspicion that file I/O is causing the issue here, like often in OsmAnnd. And with that it in deed may be a peculiarity of rhe particular hardware and/ or Android 12. And we are rather deep in google's 'secrets' of modifying file access mechanisms over the years. 😞

Chances are that a build with 'All files access' permission would solve the issue (,#16059) which would allow using any path (like just sd-card/osmand) as the data folder, and Android would then usd a different I/O access mechsnism, I think.

There may be other workarounds like better chaching, or using transparent zipping (#10653).

**Comment by sonora** [comment]
PS: Perhaps using the Download folder (or on some devices subfolder of it) as OsmAnd's data folder makes a difference, not sure. If yes, it would mean that the media access I/O on some Android devices is just slow (and probably noone cares, beceause for listening to a song or watching a video it's still good enough...)

There are also marginal chances of cheap SD cards limiting this, or a device having a slow sd-card reader. But foremost I suspect it's the way Android 12 handles file I/O, and/or perhaps media access.

**Comment by Rowin63** [comment]
I will move the data folder to "downloads/...) and give it a try. But meanwhile I'm getting closer and closer to the last A12 update (and your research pretty confirms that)...

**Comment by Rowin63** [comment]
Moved the data folder to download/net.osmand.plus/files
Yes, it's faster than the media folder is (ca 15sec to display the track content), but zoom&pan still is... no fun.

**Comment by Rowin63** [comment]
Also "data" is (and "obb" probably will be too) quick. But unfortunately no synching (anymore) possible in this paths...

**Comment by sonora** [comment]
Yes, my understanding is google also uses yet another file I/O mechanism for the Download folder, amd also that is severely (performance)-limited vs. what the hardware could actualy deliver.

The only sure-fire way to mitigate this reliably seems to use an OsmAnd build which simply gives the app the 'All files' access. Something google wants to restrict to file manager type apps ever since Android 11. Hence my attempt to convince Victor registering OsmAnd as a GPX File Manager app (#16059)...

If you run your own private build, it's a 1 line change in the code to try...

**Comment by Rowin63** [comment]
>OsmAnd as a GPX File Manager app (#16059)...
> 

I know there's a discussion about this... 

> If you run your own private build, it's a 1 line change in the code to try...
>

Unfortunately not. I don't have any knowledge about that.. 

But thanks anyway for your help 👍

**Comment by sonora** [comment]
Perhaps we could convince Victor to publish a '_test_' 😉 OsmAnd build with 'All files permission' on the OsmAnd download server... but I am afraid no one may then use the official 'google' build any more ... haha... 😉😘😉

**Comment by Rowin63** [comment]
I want to give you an update, that solves my problem (extremely slow access to "media" folder in Android 12) for the moment (I had to move Osmands data to "media" , because I couldn't synch anymore the "data" and "obb" path):

For some reason it is possible again, to add "obb" (but not "data") as a "user defined exception" in the synching app I like to use (Folder Sync Pro). So now I am able again to use the (quick) "obb"-(multiuser 1)-path as Osmands Data Folder, though having access to Osmands data for synching 👍

**Comment by sonora** [comment]
Thanks! I will add this hint to the documentation https://osmand.net/docs/user/troubleshooting/maps-data/#maps-slowly-loading-on-android-11-12-sd-card. Do you use 'Manually specified' or one of the other options in OsmAnd's Data folder menu?

**Comment by Rowin63** [comment]
I use Osmands predefined "multiuser 1" (the internal memory, "2" would be SD card, never tried that).

**Comment by Rowin63** [comment]
The mentioned issue about deleting maps at an update did not happen anymore (to me 😉)

**Comment by sonora** [comment]
Ah ok! But then it's simply fast because you use the internal storage... So it's no real solution for users who have limited internal storage, and are looking to use an external SD card's folder readily accessible by other processes, e.g. for syncing...(?)

**Comment by Rowin63** [comment]
> Ah ok! But then it's simply fast because you use the internal storage... So it's no real solution for users who have limited internal storage, and are looking to use an external SD card's folder readily accessible by other processes, e.g. for syncing...(?)

I will give it a try later today. Just need to move the app over to "2". Will tell you then 😉

**Comment by Rowin63** [comment]
Moved the data folder to "multiuser 2" (obb/SD, used Osmands predefined setting).
Is as slow as media path. Tracks menu content takes a minute to show up, zoom&pan (but in fact everything you do in the app...) with tracks on display is unusable slow.
I'm happy that internal obb at least runs fine 👍

**Comment by sonora** [comment]
Yeah, expected that. It's Android new mechanisms to limit SD card access to the protected (hidden) file paths, or to certain library mechanisms like Media or Downloads, which uses slow processes.

So the 'All files' permission remains the only known mechanism so far to facilitate using a 'generally visible' folder on the external SD card with the max performance the hardware can deliver.

**Comment by vshcherb** [comment]
Needs to be retested as there were improvements recently

**Comment by vshcherb** [comment]
@Zirochkabila Our threshold tracks with 1M points in total should work reasonably well on decent device. Though we need to find the boundary

**Comment by sonora** [comment]
Things have improved big time in recent weeks, although there is still a noticeable penalty of using v2 with many tracks vs. v1.

For testing, after importing many tracks it is advisable to use a build with https://github.com/osmandapp/OsmAnd/pull/24257, or else you will test the sluggishness due to GPX db update with 'nearest city', which will by far exceed the sluggishness we are talking about in this issue 

 - [ ] The secondary issue mentioned here "_configure map > tracks is terribly slow too_" is not fixed at all. In my setup it can in deed take up to 8 seconds to enter the 'Visible maps' screen, both from the 'My Places' screen or from ' Configure map > Tracks'.

**Comment by Zirochkabila** [comment]
# Testing 1

> OsmAnd~ 5.3.0#6954m, Released: 2026-01-16

@vshcherb Noticeable freezes began to occur when displaying tracks with a total of **360k** points. 
When attempting to display tracks with **450k** points, a failure occurs - it is impossible to check the rendering performance (there is a suspicion that **500k** points is the limit for an average device - `Qualcomm Snapdragon 732G (8 nm)`, `Adreno 618`, `6 GB RAM`).

[exception (38).log](https://github.com/user-attachments/files/24672881/exception.38.log)
[exception (39).log](https://github.com/user-attachments/files/24672882/exception.39.log)
[exception (40).log](https://github.com/user-attachments/files/24672883/exception.40.log)

Import file > Display 3 tracks (3 tracks = 90k points)

[540k points_Export_2026-01-16.osf.zip](https://github.com/user-attachments/files/24672888/Export_2026-01-16.osf.zip)




<details><summary>Detsils</summary>
<table style="width:100%">
  <tr>
    <th>90</th>
    <th>180</th>
    <th>270</th>
    <th>360</th>
    <th>450 = Crash</th>
  </tr>
  <tr>
    <td><video source src="https://github.com/user-attachments/assets/5bc4b52e-f34c-4699-954c-939da2627bc3"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/d24b2ad4-c974-45dd-bd38-7d732e9faa7a"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/9dc2c592-0170-4bca-8f32-47e7758744a8"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/2905700d-d366-4fde-926b-f0643bea5bb9"  width="300" height="auto"/></td>
    <td><video source src="https://github.com/user-attachments/assets/3ce1933f-113a-4598-9579-22ea01423bf0"  width="300" height="auto"/></td>
  </tr>
</table> 
</details>

**Comment by vshcherb** [comment]
1. We need to test performance again and find bottleneck. Looks like we can only support without FPS going down 100K points where our ultimate goal 1M points is far away and at least 300-500k we should be able to support
2. Extend UI to monitor number of points in dev mode

**Comment by Zirochkabila** [comment]
### Testing iOS

> OsmAnd Maps 5320, released 16.01.26

After displaying more than **150** tracks (**170** in the video) - **visible freezes begin.**
When displaying more than **220**, it is impossible to use the application.



<details><summary>Details - 170</summary>
<video src="https://github.com/user-attachments/assets/918f0215-46a2-4f25-948c-43a02440801c"></video>
</details> 

[265 tracks Export_19-01-26.osf.zip](https://github.com/user-attachments/files/24718893/Export_19-01-26.osf.zip)

**Comment by Zirochkabila** [comment]
- [ ] [Entering the 'Visible tracks' screen is excessively slow](https://github.com/osmandapp/OsmAnd/issues/24554#issuecomment-3980535575)

**Comment by vshcherb** [comment]
Reproduced on master (#26185 regression): with 100 visible tracks (2000 points each), `GPXLayer.onPrepareBufferImage` takes **104–115 ms per frame on the UI thread** while panning (OpenGL, debug build, API 35 emulator), for tracks inside and outside the viewport alike.

Cause: the layer calls `GpxDbHelper.getItem()` 3–5 times per visible track per frame (segments, splits, split params). Since 8325deeba2 (#25681, master only) every call does `file.exists()` plus `lastModified()` and resolves `getGpxDir()` via okio before anything else, so at 100 tracks that's hundreds of `stat()` calls per frame. On SD card / media storage (the original report here) each one is far slower. `isGpxFileSelected` also linearly scans the selected list per track, O(n²) per frame. A separate 7% of CPU goes to `LocalePreferences.getTemperatureUnit()` (ICU), which the weather layer asks for on every frame.

A fix that resolves the DB items once per track per frame, makes `getItem` do at most one stat, and caches the system temperature unit brings it to **5–6 ms per frame**, with 2–3× more frames rendered during the same swipes (interleaved A/B, two passes each).

Not addressed: tracks that leave the viewport are still removed from the renderer and rebuilt when they come back (~6% CPU).

(gen by AI)

**Comment by vshcherb** [comment]
Yes there was a regression from 10-15 ms up to 100 ms now is fixed to 2 ms

**Comment by sonora** [comment]
Fix works. But two things to check here, I guess:

- Replace `selection != trackDbItemsSelection` checks object identity, not content equality. May be a problem for the cache?

- `trackDbItems` is not thread‑safe, I think. Could be an issue?
